### PR TITLE
Allow cleanup on closing more-info

### PR DESCRIPTION
**Pull request by andrey-git**
Remove camera's streaming source from camera's 'more info' when more-info is closed.

### CODE REVIEW

File: src/more-infos/more-info-content.html [hunk 1]
```diff
@@ -35,11 +35,14 @@
...
 
   stateObjChanged: function (stateObj) {
-    if (!stateObj) return;
-
-    window.hassUtil.dynamicContentUpdater(
-      this, 'MORE-INFO-' + window.hassUtil.stateMoreInfoType(stateObj).toUpperCase(),
-      { hass: this.hass, stateObj: stateObj });
+    if (!stateObj) {
+      window.hassUtil.dynamicContentUpdater(
+        this, null, { cleanup: true });
```
Comment (balloob) [inline src/more-infos/more-info-content.html]: Can you rename it to `isVisible`, which is more in line with what it represents

File: src/more-infos/more-info-camera.html [hunk 2]
```diff
@@ -29,21 +29,26 @@
...
   imageLoaded: function () {
     this.fire('iron-resize');
   },
 
-  computeCameraImageUrl: function (hass, stateObj) {
+  computeCameraImageUrl: function (hass, stateObj, isVisible) {
     if (hass.demo) {
       return '/demo/webcam.jpg';
-    } else if (stateObj) {
+    } else if (stateObj && isVisible) {
```
Comment (balloob) [inline src/more-infos/more-info-camera.html]: I'm wondering now, I thought that this logic of `isVisible` was provided by the fact that stateObj is set to null when the more info dialog is hidden, is that no longer the case?

Comment (andrey-git) [inline src/more-infos/more-info-camera.html]: The call is "window.hassUtil.dynamicContentUpdater(this, null, { isVisible: false })"
And dynamicContentUpdater doesn't clear attributes which are not passed, so just isVisible is changed to false but, stateObj keeps its previous value.

At first I though just to pass stateObj:null instead of adding a new attribute, but some other classes expect stateObj to be non null.

File: src/more-infos/more-info-camera.html [hunk 3]
```diff
@@ -29,21 +29,26 @@
...
       return '/api/camera_proxy_stream/' + stateObj.entityId +
              '?token=' + stateObj.attributes.access_token;
     }
-    // Return an empty image if no stateObj (= dialog not open)
+    // Return an empty image if no stateObj (= dialog not open) or in cleanup mode.
```
Comment (balloob) [inline src/more-infos/more-info-camera.html]: This comment is no longer correct, stateObj will not be null.

File: src/more-infos/more-info-content.html [hunk 4]
```diff
@@ -35,11 +35,14 @@
...
 
   stateObjChanged: function (stateObj) {
-    if (!stateObj) return;
-
-    window.hassUtil.dynamicContentUpdater(
-      this, 'MORE-INFO-' + window.hassUtil.stateMoreInfoType(stateObj).toUpperCase(),
-      { hass: this.hass, stateObj: stateObj });
+    if (!stateObj) {
+      window.hassUtil.dynamicContentUpdater(
+        this, null, { isVisible: false });
```
Comment (balloob) [inline src/more-infos/more-info-content.html]: by setting it to null, if we close a more info light and open it again, it has to re-create all the DOM. The old logic was faster. 

Also, no need to pass attributes if you're not going to render anything?

I think that it might make more sense to just check the lastChild, and if it exists set `lastChild.isVisible = false`

File: src/util/hass-util.html [hunk 5]
```diff
@@ -47,16 +47,20 @@
...
 
+// Update root's child element to be newElementTag replacing another existing child if any.
+// Copy attributes into the child element.
+// If newElementTag is null just update the existing child element if any.
 window.hassUtil.dynamicContentUpdater = function (root, newElementTag, attributes) {
   var rootEl = Polymer.dom(root);
   var customEl;
 
-  if (rootEl.lastChild && rootEl.lastChild.tagName === newElementTag) {
+  if (rootEl.lastChild && (rootEl.lastChild.tagName === newElementTag || !newElementTag)) {
```
Comment (balloob) [inline src/util/hass-util.html]: Not sure if I see what is happening here? We should only assign customEl to lastChild if it is the same tag.

File: src/util/hass-util.html [hunk 6]
```diff
@@ -47,16 +47,20 @@
...
     customEl = rootEl.lastChild;
   } else {
     if (rootEl.lastChild) {
       rootEl.removeChild(rootEl.lastChild);
     }
+    if (!newElementTag) return;
```
Comment (balloob) [inline src/util/hass-util.html]: Why return here? If we just want to remove the lastChild if this is null, we should move that to the top of the function. Now the logic gets hard to follow.

File: src/more-infos/more-info-camera.html (hunk 3 above)
Comment (andrey-git) [inline src/more-infos/more-info-camera.html]: stateObj can still be null is some flow I didn't change. (i.e. not from dynamicContentUpdater)
In that case camera was using the fallback image and will still do so after my change.

File: src/more-infos/more-info-content.html (hunk 4 above)
Comment (andrey-git) [inline src/more-infos/more-info-content.html]: Before my change closing more-info would just hide the element. At least in chrome camera stream  would still continue to run.

I tried to remove (without resetting the URL) the more-info-camera from the DOM, but the stream would still run.

I refactored this to use different functions for the null and non-null cases.

Regarding speed:
1) opening more-info-light [no change]
2) closing it. [sets isVisible to false, light don't care but a small amount of work is done]
3) opening it again [exactly same effort. The dom is still there and received a stateObj like before my change]

File: src/util/hass-util.html (hunk 5 above)
Comment (andrey-git) [inline src/util/hass-util.html]: Refactored.

File: src/util/hass-util.html (hunk 6 above)
Comment (andrey-git) [inline src/util/hass-util.html]: Refactored

File: src/more-infos/more-info-content.html (hunk 4 above)
Comment (michaelarnauts) [inline src/more-infos/more-info-content.html]: This actually had cost me money! I wasn't aware that the camera stream kept running when the dialog was closed, and it kept on streaming on my mobile phone.

File: src/util/hass-util.html [hunk 7]
```diff
@@ -63,12 +65,23 @@
   Object.keys(attributes).forEach(function (key) {
     customEl[key] = attributes[key];
   });
+  customEl.isVisible = true;
```
Comment (balloob) [inline src/util/hass-util.html]: isVisible is not part of dynamicContentUpdater. dynamicContentUpdater only dynamically updates content, it doesn't mean that the content will be visible.

Comment (andrey-git) [inline src/util/hass-util.html]: I'm looking for a way to cancel the effect of a previous call to resetDynamicContent.

Comment (andrey-git) [inline src/util/hass-util.html]: I could add 'isVisible':false to stateObj before calling dynamicContentUpdater in more-info-content.html

What do you think?

Comment (balloob) [inline src/util/hass-util.html]: Yes, adding it to more-info-content.html sounds like the way to go. We wouldn't need `resetDynamicContent` - `dynamicContentUpdater` could just be used?

Comment (andrey-git) [inline src/util/hass-util.html]: Moved isVisible=true setting to  more-info-content.html

File: src/util/hass-util.html [hunk 8]
```diff
@@ -69,6 +71,16 @@
   }
 };
 
+// If root has lastChild, set 'isVisible' attribute of that child to false.
+window.hassUtil.resetDynamicContent = function (root) {
```
Comment (balloob) [inline src/util/hass-util.html]: I think that you should inline this logic into `more-info-content.html`, it's the only place that knows about `isVisible`. Alternative, we could decide to just remove the whole `more-info` tag from the DOM when there is no stateObj, so there won't even be a need for `isVisible`

Comment (andrey-git) [inline src/util/hass-util.html]: Removing the more-info tag is not good enough - the stream continues to run. The src must be changed to non-stream to stop it.

I'll inline resetDynamicContent into more-info-content.html

Comment (balloob) [inline src/util/hass-util.html]: When we remove the child from the DOM, the image is no longer attached to the DOM either and thus it should stop streaming? Otherwise that might be a browser bug.

Comment (andrey-git) [inline src/util/hass-util.html]: I tried it with chrome desktop and mobile.

Probably the object is not destroyed when removed. Maybe there is a reference to it somewhere in our code or in Polymer.
